### PR TITLE
register `on.exit()` hook before estimation in `tune_bayes()`

### LINKED ISSUE #340: `tune_bayes()` doesn't return a `.notes` column on error but claims that it does
I've set up a model spec below that I know will make the tuning error. Both `tune_grid()` and `tune_bayes()` claim to return an object with a `.notes` column, but only `tune_grid()` actually does; `tune_bayes()` just errors out. Not sure what the intended behaviour for `tune_bayes()` actually is in this case, but happy to help with a PR if you spell it out for me :-)

``` r
library(tidymodels)
#> ── Attaching packages ────────────────────────────────────── tidymodels 0.1.2 ──
#> ✓ broom     0.7.2      ✓ recipes   0.1.15
#> ✓ dials     0.0.9      ✓ rsample   0.0.8 
#> ✓ dplyr     1.0.2      ✓ tibble    3.0.4 
#> ✓ ggplot2   3.3.2      ✓ tidyr     1.1.2 
#> ✓ infer     0.5.3      ✓ tune      0.1.2 
#> ✓ modeldata 0.1.0      ✓ workflows 0.2.1 
#> ✓ parsnip   0.1.4      ✓ yardstick 0.0.7 
#> ✓ purrr     0.3.4
#> ── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard() masks scales::discard()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x recipes::step()  masks stats::step()
set.seed(1)

bt <- boost_tree(mode = "classification",
                 trees = tune(), tree_depth = tree_depth(range = 1:2),
                 learn_rate = tune()) %>% 
  set_engine("xgboost")

rec <- recipe(Species ~ ., data = iris)

wf <- workflow() %>% 
  add_recipe(rec) %>% 
  add_model(bt)

rs <- vfold_cv(iris, v = 4)

tg <- tune_grid(wf, rs)
#> 
#> Attaching package: 'rlang'
#> The following objects are masked from 'package:purrr':
#> 
#>     %@%, as_function, flatten, flatten_chr, flatten_dbl, flatten_int,
#>     flatten_lgl, flatten_raw, invoke, list_along, modify, prepend,
#>     splice
#> 
#> Attaching package: 'vctrs'
#> The following object is masked from 'package:tibble':
#> 
#>     data_frame
#> The following object is masked from 'package:dplyr':
#> 
#>     data_frame
#> 
#> Attaching package: 'xgboost'
#> The following object is masked from 'package:dplyr':
#> 
#>     slice
#> x Fold1: preprocessor 1/1, model 1/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 2/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 3/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 4/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 5/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 6/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 7/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 8/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 9/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold1: preprocessor 1/1, model 10/10: Error in xgb.iter.update(bst$handle, dtra...
#> x Fold2: preprocessor 1/1, model 1/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 2/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 3/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 4/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 5/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 6/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 7/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 8/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 9/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold2: preprocessor 1/1, model 10/10: Error in xgb.iter.update(bst$handle, dtra...
#> x Fold3: preprocessor 1/1, model 1/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 2/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 3/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 4/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 5/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 6/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 7/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 8/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 9/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold3: preprocessor 1/1, model 10/10: Error in xgb.iter.update(bst$handle, dtra...
#> x Fold4: preprocessor 1/1, model 1/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 2/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 3/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 4/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 5/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 6/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 7/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 8/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 9/10: Error in xgb.iter.update(bst$handle, dtrai...
#> x Fold4: preprocessor 1/1, model 10/10: Error in xgb.iter.update(bst$handle, dtra...
#> Warning: All models failed. See the `.notes` column.
print(tg)
#> Warning: This tuning result has notes. Example notes on model fitting include:
#> preprocessor 1/1, model 10/10: Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj): Invalid Parameter format for max_depth expect int but value='integer'
#> preprocessor 1/1, model 1/10: Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj): Invalid Parameter format for max_depth expect int but value='integer'
#> preprocessor 1/1, model 7/10: Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj): Invalid Parameter format for max_depth expect int but value='integer'
#> # Tuning results
#> # 4-fold cross-validation 
#> # A tibble: 4 x 4
#>   splits           id    .metrics .notes           
#>   <list>           <chr> <list>   <list>           
#> 1 <split [112/38]> Fold1 <NULL>   <tibble [10 × 1]>
#> 2 <split [112/38]> Fold2 <NULL>   <tibble [10 × 1]>
#> 3 <split [113/37]> Fold3 <NULL>   <tibble [10 × 1]>
#> 4 <split [113/37]> Fold4 <NULL>   <tibble [10 × 1]>

tb <- tune_bayes(wf, rs)
#> x Fold1: preprocessor 1/1, model 1/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold1: preprocessor 1/1, model 2/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold1: preprocessor 1/1, model 3/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold1: preprocessor 1/1, model 4/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold1: preprocessor 1/1, model 5/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold2: preprocessor 1/1, model 1/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold2: preprocessor 1/1, model 2/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold2: preprocessor 1/1, model 3/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold2: preprocessor 1/1, model 4/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold2: preprocessor 1/1, model 5/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold3: preprocessor 1/1, model 1/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold3: preprocessor 1/1, model 2/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold3: preprocessor 1/1, model 3/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold3: preprocessor 1/1, model 4/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold3: preprocessor 1/1, model 5/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold4: preprocessor 1/1, model 1/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold4: preprocessor 1/1, model 2/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold4: preprocessor 1/1, model 3/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold4: preprocessor 1/1, model 4/5: Error in xgb.iter.update(bst$handle, dtrain...
#> x Fold4: preprocessor 1/1, model 5/5: Error in xgb.iter.update(bst$handle, dtrain...
#> Warning: All models failed. See the `.notes` column.
#> Error: All of the models failed. See the .notes column.
print(tb)
#> Error in print(tb): object 'tb' not found
```

<sup>Created on 2020-12-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #340.

Moves the `on.exit()` before `estimate_tune_results` so that `tune_bayes()` returns initial results, as the current error says it does.

### CODE REVIEW

**Review by DavisVaughan — APPROVED** [review]

**Comment by ararifuddinr** [comment]
I want to tune the xgboost model using bayesian optimization by tidymodels but when defining the range of hyperparameter values ​​there is a problem. Can u help me Sir??
I want specific hyperparameter combinations so i use expand.grid() function to define hyperparameter values

library(visdat)
library(tidyverse)
library(tidymodels)
#> Registered S3 method overwritten by 'tune':
#>   method                   from   
#>   required_pkgs.model_spec parsnip
#> Warning: package 'dials' was built under R version 4.1.3
#> Warning: package 'parsnip' was built under R version 4.1.3
library(patchwork)
library(readxl)
library(ranger)
library(reprex)
#> Warning: package 'reprex' was built under R version 4.1.3

data <- mtcars

data$cyl <- as.factor(data$cyl)
data$vs <- as.factor(data$vs)
data$am <- as.factor(data$am)
data$gear <- as.factor(data$gear)
data$carb <- as.factor(data$carb)

set.seed(123)
data.training <- data

xgboost_recipe <- 
  recipe(formula = mpg ~ ., data = data.training) %>% 
  step_novel(all_nominal(), -all_outcomes()) %>%
  step_dummy(all_nominal(), -all_outcomes(), one_hot = TRUE) %>% 
  step_zv(all_predictors()) 

xgboost_spec <- 
  boost_tree(trees = tune(), min_n = tune(), tree_depth = tune(), learn_rate = tune(), 
             loss_reduction = tune(), sample_size = tune(), mtry = tune()) %>% 
  set_mode("regression") %>% 
  set_engine("xgboost") 

# define hyperparameter range

params <- expand.grid(
  trees = seq(500,1000,by=200),
  min_n = seq(10,40, by=10),
  tree_depth = seq(6L, 10L, by=2L),
  learn_rate = seq(0.1,1, by=0.5),
  loss_reduction = seq(0.1,1, by=0.5),
  sample_prop = seq(0.1,1, by=0.5), mtry=seq(2,7, by=1))

xgboost_workflow <- 
  workflow() %>% 
  add_recipe(xgboost_recipe) %>% 
  add_model(xgboost_spec) 

# resampling

set.seed(123)
resampling <- vfold_cv(data = data.training, v = 10, strata = mpg)
#> Warning: The number of observations in each quantile is below the recommended threshold of 20.
#> * Stratification will use 1 breaks instead.
#> Warning: Too little data to stratify.
#> * Resampling will be unstratified.


# Tune use bayesian optimization

doParallel::registerDoParallel()

set.seed(456) 
res <-
  tune_bayes(xgboost_workflow,
             iter = 6,
             resamples = resampling, 
             param_info = params, 
             metrics = metric_set(mae),
             control = control_bayes(verbose = TRUE, 
                                     save_pred = TRUE)
  )
#> ! There are 1728 tuning parameters and 5 grid points were requested. This is
#>   likely to cause numerical issues in the first few search iterations.
#> Error in UseMethod("grid_latin_hypercube"): no applicable method for 'grid_latin_hypercube' applied to an object of class "data.frame"

**Comment by juliasilge** [comment]
@ararifuddinr We definitely want you to get the help you need! It's not that useful to post a question like this on a merged PR, unfortunately. Could you update your reproducible example to use the [reprex package](http://reprex.tidyverse.org)? Using reprex makes it easier to see both the input and output, and for people to re-run the code in a local session. Once you have a reprex, I recommend posting on [RStudio Community](https://rstd.io/tidymodels-community), which is a great forum for getting help with these kinds of modeling questions.
